ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParsePayload.java

```diff
@@ -75,132 +76,133 @@ public static ParsePayload of(String schemasLocation) {
   private transient PipelineMetadataStore metadataStore;
   private transient CRC32 crc32;
 
+  final TupleTag<PubsubMessage> outputTag = new TupleTag<>() {
+  };
+  final TupleTag<PubsubMessage> failureTag = new TupleTag<>() {
+  };
+
   private ParsePayload(String schemasLocation) {
     this.schemasLocation = schemasLocation;
   }
 
   @Override
   public Result<PCollection<PubsubMessage>, PubsubMessage> expand(
       PCollection<PubsubMessage> messages) {
-    return messages.apply(FlatMapElements.into(TypeDescriptor.of(PubsubMessage.class)) //
-        .via(new Fn()) //
-        .exceptionsInto(TypeDescriptor.of(PubsubMessage.class)) //
-        .exceptionsVia((WithFailures.ExceptionElement<PubsubMessage> ee) -> {
-          try {
-            throw ee.exception();
-          } catch (IOException | SchemaNotFoundException | ValidationException
-              | MessageScrubberException | DeprecatedMessageException e) {
-            return FailureMessage.of(ParsePayload.class.getSimpleName(), //
-                ee.element(), //
-                ee.exception());
-          }
-        }));
+    PCollectionTuple parsed = messages
+        .apply(ParDo.of(new Fn()).withOutputTags(outputTag, TupleTagList.of(failureTag)));
+    return WithFailures.Result.of(parsed.get(outputTag), parsed.get(failureTag));
   }
 
-  private class Fn implements ProcessFunction<PubsubMessage, Iterable<PubsubMessage>> {
+  private class Fn extends DoFn<PubsubMessage, PubsubMessage> {
 
-    @Override
-    public Iterable<PubsubMessage> apply(PubsubMessage message) throws IOException {
-      message = PubsubConstraints.ensureNonNull(message);
-      Map<String, String> attributes = new HashMap<>(message.getAttributeMap());
+    @Setup
+    public void setup() {
+      schemaStore = SchemaStoreSingletonFactory.getJsonSchemaStore(schemasLocation);
+      metadataStore = SchemaStoreSingletonFactory.getPipelineMetadataStore(schemasLocation);
+      validator = new JsonValidator();
+    }
 
-      if (schemaStore == null) {
-        schemaStore = JSONSchemaStore.of(schemasLocation, BeamFileInputStream::open);
-      }
+    @ProcessElement
+    public void processElement(@Element PubsubMessage message, MultiOutputReceiver out) {
```

Comment:
This is mostly the same code that was in `apply()` method before, the only differences are:
* schemastore and validator initializations are moved to `setup()`
* exception handling is moved here from `expand()` above since now we explicitly use `ParDo`
* instead of `return`ing results we write them to output receiver